src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -0,0 +1,346 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+import is.hail.annotations.MemoryBuffer
+import is.hail.asm4s.FunctionBuilder.methodNodeToGrowable
+import is.hail.asm4s._
+import is.hail.expr
+import is.hail.expr.{TInt32, TInt64, TArray, TContainer, TStruct, TFloat32, TFloat64, TBoolean}
+import is.hail.annotations.StagedRegionValueBuilder
+import scala.collection.generic.Growable
+
+import org.objectweb.asm.Opcodes._
+import org.objectweb.asm.Type
+import org.objectweb.asm.tree._
+
+import scala.reflect.classTag
+import scala.reflect.ClassTag
+
+object Compile {
+  private def dummyValue(t: expr.Type): Code[_] = t match {
+    case TBoolean => false
+    case TInt32 => 0
+    case TInt64 => 0L
+    case TFloat32 => 0.0f
+    case TFloat64 => 0.0
+    case _ => 0L // reference types
+  }
+  private def typeToTypeInfo(t: expr.Type): TypeInfo[_] = t match {
+    case TInt32 => typeInfo[Int]
+    case TInt64 => typeInfo[Long]
+    case TFloat32 => typeInfo[Float]
+    case TFloat64 => typeInfo[Double]
+    case TBoolean => typeInfo[Boolean]
+    case _ => typeInfo[Long] // reference types
+  }
+
+  type E = Env[(TypeInfo[_], Code[Boolean], Code[_])]
+
+  def apply(ir: IR, fb: FunctionBuilder[_]) {
+    apply(ir, fb, new Env())
+  }
+
+  def apply(ir: IR, fb: FunctionBuilder[_], env: E) {
+    fb.emit(expression(ir, fb, env, new StagedBitSet(fb))._2)
+  }
+
+  private def present(x: Code[_]) = (const(false), x)
+  private def coerce[T](x: Code[_]): Code[T] = x.asInstanceOf[Code[T]]
+  private def tcoerce[T <: expr.Type](x: expr.Type): T = x.asInstanceOf[T]
+
+  def expression(ir: IR, fb: FunctionBuilder[_], env: E, mb: StagedBitSet): (Code[Boolean], Code[_]) = {
```

Comment:
I'm not sure I get the name `expression`.  Compile?  Emit?

What's the invariant on what gets emitted directly to `fb` vs the return type here?

Reply:
`expression` is a relic of the distinction between expressions and statements. I'll rename it to compile.

There's not really any invariant. Since `Code` cannot express snippets that need the `FunctionBuilder`, I need to sometimes emit directly to the `fb`. I didn't go the monad route because that seems to suck in Scala from past experiences.

I've been thinking it would be cleaner to never return `Code` and just `emit` instructions directly.

Reply:
I guess the trouble with just using `emit` is that I can't use a bunch of our `Code` abstractions like `mux` and `whileLoop`. I'll have to write new ones that take `() => ()` thunks with the side effect of emitting instructions.

Reply:
> There's not really any invariant.

I'm going to have a hard time reasoning about the correctness of your method then.  Does the value code need to be evaluating once and only once to get the effect of the IR?  And it can contain arbitrary mutation?  And it must be emitted after whatever was emitted to the function builder by the call to `expression`?  (These are certainly invariants, and I think they're yours, I'm just not 100% sure.)  What about the missingness code?  Can I run it more than once?  Do I have to run it at all?  Can it perform mutation?

Here is the invariant I would use: calling `expression` emits the code that evaluates the IR into the FunctionBuilder.  The return value is a pair of codes which return the result of the IR, are functional, and can be evaluated 0, 1 or more times.  (Typically, they will be references to locals holding the result of evaluating the IR.)

The opposite extreme is also OK: nothing is emitted to the fb (you just pass it in for services like variable creation) and all the code generated is returned.  You're still going to have to say something about the relationship between the missing and value `Code`s, seems like to me.

 > I guess the trouble with just using emit is that I can't use a bunch of our Code abstractions like mux and whileLoop.

Sure you can, you just need to emit them at the end.

Thoughts?